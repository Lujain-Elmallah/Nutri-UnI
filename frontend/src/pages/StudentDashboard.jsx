import { useEffect, useState } from "react";
import { getWeekMenu } from "../api/meals";
import MenuTable from "@/components/MenuTable";
import Header from "@/components/Header"; 

function StudentDashboard() {
  const [menu, setMenu] = useState([]);
  const [filteredMenu, setFilteredMenu] = useState([]);
  const [loading, setLoading] = useState(true);
  const [query, setQuery] = useState("");
  const [selectedCounter, setSelectedCounter] = useState(null);
  const [error, setError] = useState("");

  useEffect(() => {
    const fetchMenu = async () => {
      setLoading(true);
      try {
        const res = await getWeekMenu();
        setMenu(res.data);
        setFilteredMenu(res.data);
        setError("");
      } catch (err) {
        console.error(err);
        setError("Unable to load meals at counter. Please try again later.");
      }
      setLoading(false);
    };

    fetchMenu();
  }, []);

  const handleSearch = (q) => {
    setQuery(q);
    applyFilters(q, selectedCounter);
  };

  const handleCounterFilter = (counter) => {
    const newCounter = counter === selectedCounter ? null : counter;
    setSelectedCounter(newCounter);
    applyFilters(query, newCounter);
  };

  const applyFilters = (q, counter) => {
    let filtered = [...menu];

    if (counter) {
      filtered = filtered.filter((item) => item.counter?.toLowerCase() === counter.toLowerCase());
    }

    if (q.trim()) {
      const keyword = q.toLowerCase();
      filtered = filtered.filter(
        (item) =>
          item.item_name?.toLowerCase().includes(keyword) ||
          (item.allergens && item.allergens.join(" ").toLowerCase().includes(keyword)) ||
          (item.vegan && keyword === "vegan") ||
          (item.veg && keyword === "vegetarian") ||
          (item.gluten_free && keyword === "gluten free")
      );
    }

    setFilteredMenu(filtered);
  };

  const uniqueCounters = [...new Set(menu.map((item) => item.counter))];

  return (
    <>
      <Header showItemsTab={false} /> 

      <div className="pt-20 px-4 md:px-6 lg:px-8">
        <div className="w-full max-w-[1600px] mx-auto space-y-6">
          <h1 className="text-2xl font-header-primary text-[#303030]">Student Dashboard</h1>

          <div className="flex flex-col md:flex-row gap-4 items-center">
            <input
              type="text"
              placeholder="Search meals by name, ingredient, or dietary category"
              value={query}
              onChange={(e) => handleSearch(e.target.value)}
              className="w-full max-w-md px-4 py-2 border rounded-xl shadow"
            />
          </div>

          <div className="flex flex-wrap gap-3 mt-4">
            {uniqueCounters.map((counter) => (
              <button
                key={counter}
                onClick={() => handleCounterFilter(counter)}
                className={`px-4 py-2 rounded-xl shadow ${
                  selectedCounter === counter ? "bg-[#008b9e] text-white" : "bg-gray-200"
                }`}
              >
                {counter}
              </button>
            ))}
          </div>

          {error && (
            <div className="text-red-600 mt-4 text-lg font-medium">{error}</div>
          )}

          {!loading && filteredMenu.length === 0 && !error && (
            <div className="text-center text-lg text-gray-500 mt-4">
              {selectedCounter
                ? "No meals available at this counter today."
                : "No meals found. Try a different keyword."}
            </div>
          )}

          <MenuTable menuItems={filteredMenu} loading={loading} hideAdminControls /> 
        </div>
      </div>
    </>
  );
}

export default StudentDashboard;
