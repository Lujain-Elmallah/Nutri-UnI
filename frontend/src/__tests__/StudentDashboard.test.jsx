import { render, screen } from "@testing-library/react";
import StudentDashboard from "../pages/StudentDashboard";
import { MemoryRouter } from "react-router-dom";
import { act } from "react";
import { vi } from "vitest";

// Mock Clerk
vi.mock("@clerk/clerk-react", async () => {
  const actual = await vi.importActual("@clerk/clerk-react");
  return {
    ...actual,
    useUser: () => ({
      user: {
        fullName: "John Doe",
        publicMetadata: { role: "student" },
      },
      isLoaded: true,
    }),
  };
});

// Mock API
vi.mock("../api/meals", () => ({
  getWeekMenu: () => Promise.resolve({
    data: [
      {
        id: "1",
        dining_hall: "D1",
        counter: "Mongolian",
        food_info: {
          item_name: "Greek Salad",
          energy: 200,
          protein: 10,
          fats: 5,
          sugar: 3,
          salt: 1,
          allergens: ["nuts"],
          veg: true,
          vegan: false,
          gluten_free: true,
        },
      },
    ],
  }),
}));

test("renders greeting for student", async () => {
  await act(async () => {
    render(
      <MemoryRouter>
        <StudentDashboard />
      </MemoryRouter>
    );
  });

  expect(screen.getByText((text) => text.includes("Hello"))).toBeInTheDocument();
});

test("displays dining halls with flexible matchers", async () => {
  await act(async () => {
    render(
      <MemoryRouter>
        <StudentDashboard />
      </MemoryRouter>
    );
  });

  expect(await screen.findByText((t) => t.toLowerCase().includes("d1"))).toBeInTheDocument();
});

test("matches snapshot", async () => {
  let container;
  await act(async () => {
    const rendered = render(
      <MemoryRouter>
        <StudentDashboard />
      </MemoryRouter>
    );
    container = rendered.container;
  });

  expect(container).toMatchSnapshot();
});

test("filters meals by counter when a counter is clicked", async () => {
    await act(async () => {
      render(
        <MemoryRouter>
          <StudentDashboard />
        </MemoryRouter>
      );
    });
  
    // Step 1: Click "D1" to reveal its counters
    const d1Button = await screen.findByRole("button", { name: /d1/i });
    expect(d1Button).toBeInTheDocument();
  
    await act(async () => {
      d1Button.click();
    });
  
    // Step 2: Now "Mongolian" counter should be visible
    const counterButton = await screen.findByRole("button", { name: /mongolian/i });
    expect(counterButton).toBeInTheDocument();
  
    await act(async () => {
      counterButton.click();
    });
  
    // Step 3: Check if filtered meal is visible
    expect(await screen.findByText(/greek salad/i)).toBeInTheDocument();
  });
  