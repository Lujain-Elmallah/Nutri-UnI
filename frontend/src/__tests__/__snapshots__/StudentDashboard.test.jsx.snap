// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <header
    class="fixed top-6 left-1/2 transform -translate-x-1/2 w-[92%] max-w-[1400px] bg-[#fff9f0] rounded-2xl shadow-lg px-6 py-4 flex items-center justify-between z-50"
  >
    <div
      class="text-2xl text-[#303030] tracking-tight"
    >
      Nutri-Uni
    </div>
    <div
      class="relative w-120 gap-60"
    >
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 flex h-9 w-full min-w-0 px-3 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-white border border-[#e5dcdc] rounded-[10px] pl-10 py-2"
        data-slot="input"
        placeholder="Search meals by name, ingredient, or dietary category"
        value=""
      />
      <svg
        class="lucide lucide-search absolute left-3 top-1/2 transform -translate-y-1/2 h-5 w-5 text-gray-400"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
    </div>
    <div
      class="flex items-center gap-8"
    >
      <nav
        class="flex gap-6 font-medium text-sm"
      >
        <button
          class="hover:underline text-[#303030]"
        >
          Week's Menu
        </button>
      </nav>
      <svg
        class="lucide lucide-circle-user h-6 w-6 text-[#F9A826]"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <circle
          cx="12"
          cy="10"
          r="3"
        />
        <path
          d="M7 20.662V19a2 2 0 0 1 2-2h6a2 2 0 0 1 2 2v1.662"
        />
      </svg>
    </div>
  </header>
  <div
    class="pt-20 px-4 md:px-6 lg:px-8"
  >
    <div
      class="w-full max-w-[1600px] mx-auto space-y-6"
    >
      <h1
        class="text-2xl font-header-primary text-[#303030]"
      >
        Hello 
        there
        ,
      </h1>
      <div
        class="space-y-8"
      >
        <h2
          class="text-xl font-medium text-left"
        >
          Please select your dining hall:
        </h2>
        <div
          class="flex flex-col gap-4"
        >
          <button
            class="w-full p-6 rounded-xl shadow-sm bg-[#ebf6f7] hover:bg-[#bde6ea] transition-colors text-center"
          >
            <div
              class="text-xl font-medium"
            >
              D1
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
